src/Disks/DiskWebServer.cpp

```diff
@@ -148,181 +133,106 @@ class ReadBufferFromWebServer final : public ReadIndirectBufferFromRemoteFS<Read
-    bool can_throw = CurrentThread::isInitialized() && CurrentThread::get().getQueryContext();
-
-    try
-    {
-        if (!metadata.tables_data.count(uuid))
-            metadata.initialize(uri, settings->files_prefix, uuid, getContext());
-    }
-    catch (const Poco::Exception &)
+    std::vector<fs::path> dir_file_paths;
+    if (files.find(path) == files.end())
     {
-        const auto message = getCurrentExceptionMessage(false);
-        if (can_throw)
+        try
         {
-            throw Exception(ErrorCodes::NETWORK_ERROR, "Cannot load disk metadata. Error: {}", message);
+            initialize(fs::path(url) / path);
         }
+        catch (...)
+        {
+            const auto message = getCurrentExceptionMessage(false);
+            bool can_throw = CurrentThread::isInitialized() && CurrentThread::get().getQueryContext();
+            if (can_throw)
+                throw Exception(ErrorCodes::NETWORK_ERROR, "Cannot load disk metadata. Error: {}", message);
 
-        LOG_TRACE(&Poco::Logger::get("DiskWeb"), "Cannot load disk metadata. Error: {}", message);
-        /// Empty iterator.
-        return std::make_unique<DiskWebDirectoryIterator<RootDirectoryListing>>(metadata.tables_data[""], path);
+            LOG_TRACE(&Poco::Logger::get("DiskWeb"), "Cannot load disk metadata. Error: {}", message);
+            return std::make_unique<DiskWebServerDirectoryIterator>(std::move(dir_file_paths));
+        }
     }
 
-    String directory_name;
-    if (RE2::FullMatch(path, MATCH_DIRECTORY_PATTERN, &uuid, &directory_name))
-    {
-        if (metadata.tables_data[uuid].contains(directory_name))
-            return std::make_unique<DiskWebDirectoryIterator<DirectoryListing>>(metadata.tables_data[uuid][directory_name], path);
-        if (can_throw)
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Directory {} does not exist. (uuid: {})", directory_name, uuid);
-        return std::make_unique<DiskWebDirectoryIterator<RootDirectoryListing>>(metadata.tables_data[""], path); /// Empty directory.
-    }
+    if (files.find(path) == files.end())
+        throw Exception("Directory '" + path + "' does not exist", ErrorCodes::DIRECTORY_DOESNT_EXIST);
+
+    for (const auto & file : files)
+        if (parentPath(file.first) == path)
+            dir_file_paths.emplace_back(file.first);
 
-    return std::make_unique<DiskWebDirectoryIterator<RootDirectoryListing>>(metadata.tables_data[uuid], path);
+    LOG_TRACE(log, "Iterate directory {} with {} files", path, dir_file_paths.size());
+    return std::make_unique<DiskWebServerDirectoryIterator>(std::move(dir_file_paths));
 }
 
 
 size_t DiskWebServer::getFileSize(const String & path) const
 {
-    File file;
-    if (!findFileInMetadata(path, file))
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "File {} not found", path);
-    return file.size;
+    auto iter = files.find(path);
+    if (iter == files.end())
+        return false;
```

Comment:
Should it return `0`? Maybe even throw exception?
```suggestion
        return 0;
```

Reply:
ups, thanks!
we do need to throw here.